types/validator_set.go

```diff
@@ -366,6 +385,202 @@ func (vals *ValidatorSet) Iterate(fn func(index int, val *Validator) bool) {
+
+	// Scan changes by address and append valid validators to updates or removal
+	for _, valUpdate := range changes {
+		if bytes.Equal(valUpdate.Address, prevAddr) {
+			err := fmt.Errorf("duplicate entry %v in changes list %v", valUpdate, changes)
+			return nil, nil, err
+		}
+		if valUpdate.VotingPower < 0 {
+			err := fmt.Errorf("voting power can't be negative %v", valUpdate)
+			return nil, nil, err
+		}
+		if valUpdate.VotingPower == 0 {
+			// Add valUpdate to removals
+			removals = append(removals, valUpdate)
+		} else {
+			// Add valUpdate to updates
+			updates = append(updates, valUpdate)
+		}
+		prevAddr = valUpdate.Address
+	}
+	return updates, removals, nil
+}
+
+// Verifies a list of updates against a validator set, making sure the allowed
+// total voting power would not exceed if these updates would be applied to the set.
+// Computes the proposer priority for the validators not present in the set.
+// Returns error if first step fails.
+// updates parameter must be a list of unique validators to be added or updated.
+func verifyUpdatesAndComputeNewPriorities(updates []*Validator, vals *ValidatorSet) error {
+
+	// Scan the updates, compute new total voting power, check for overflow
+	updatedVotingPower := vals.TotalVotingPower()
+
+	for _, valUpdate := range updates {
+		address := valUpdate.Address
+		_, val := vals.GetByAddress(address)
+		if val == nil { // add
+			updatedVotingPower +=  valUpdate.VotingPower
+		} else { // update
+			updatedVotingPower += valUpdate.VotingPower - val.VotingPower
+		}
+
+		overflow := updatedVotingPower > MaxTotalVotingPower || updatedVotingPower < 0
+		if overflow {
+			err := fmt.Errorf(
+				"failed to add/update validator %v. total voting power would exceed the max allowed %v",
+				valUpdate,
+				MaxTotalVotingPower)
+			return err
+		}
+	}
+	// Loop again and update the proposerPriority for newly added and updated validators
+	for _, valUpdate := range updates {
+		address := valUpdate.Address
+		_, val := vals.GetByAddress(address)
+		if val == nil {
+			// add val
+			// Set ProposerPriority to -C*totalVotingPower (with C ~= 1.125) to make sure validators can't
+			// un-bond and then re-bond to reset their (potentially previously negative) ProposerPriority to zero.
+			//
+			// Contract: updatedVotingPower < MaxTotalVotingPower to ensure ProposerPriority does
+			// not exceed the bounds of int64.
+			//
+			// Compute ProposerPriority = -1.125*totalVotingPower == -(updatedVotingPower + (updatedVotingPower >> 3)).
+			valUpdate.ProposerPriority = -(updatedVotingPower + (updatedVotingPower >> 3))
+		} else {
+			valUpdate.ProposerPriority = val.ProposerPriority
+		}
+	}
+
+	return nil
+}
+
+// Merges the vals' validator list with the updates list.
+// When two elements with same address are seen, the one from updates is selected.
+// Expects updates to be a list of updates sorted by address with no duplicates,
+// and to have been validated with validateUpdates().
```

Comment:
I can't find `validateUpdates`. This probably refers to `verifyUpdatesAndComputeNewPriorities`?